doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+- with `_keyData` and `_keyName` allocated: 96 bytes.
+- with `_keyData` and `_keyName` `NULL`: 32 bytes.
+
+An empty key of the current implementation has 64 bytes.
+
+A copied, non-modified key with the data structure above does always have 32 bytes.
+A copied, non-modified key of the current implementation has at least 64 bytes (for an empty key), but in reality much more as the name and the data are also copied.
+
+As an example, let's take `user:/hosts/ipv6/example.com = 2606:2800:220:1:248:1893:25c8:1946`.
+
+For a single key, it would take for the COW implementation:
+
+`sizeof(_Key) + sizeof(_KeyName) + sizeof(_KeyData) + name + 1 /* \0 */ + ukey + value + 1 /* \0 */` =
+32 + 40 + 24 + 28 + 1  + 25 + 34 + 1 = 185 bytes.
+
+For two keys (one original and one copy), it would take for the COW implementation:
+
+`sizeof(_Key)*2 + sizeof(_KeyName) + sizeof(_KeyData) + name + 1 /* \0 */ + ukey + value + 1 /* \0 */` =
+32*2 + 40 + 24 + 28 + 1  + 25 + 34 + 1 = 217 bytes.
+
+For a single key, it would take the current implementation:
+
+`sizeof(_Key) + name + 1 + ukey + value + 1` =
+64 + 28 + 1 + 25 + 34 + 1 = 153 bytes
+
+For two keys, the current implementation would take:
+
+`(sizeof(_Key) + name + 1 + ukey + value + 1) * 2` =
+(64 + 28 + 1 + 25 + 34 + 1) * 2 = 306 bytes.
+
+#### Changes to `KeySet`
+
+For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.
+This includes the array itself, the sizes and the hashmap.
+
+Why don't we just add the number of references to the original `KeySet`?
+
+- If we delete a copied KeySet, we don't know which KeySet is the original, so we couldn't decrement the counter. 
+  This could be dealt with storing a pointer to the original KeySet.
+- If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work
+- In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change). 
+  This is unexpected behaviour.
+
+```c
+struct _KeySetData {
+    struct _Key ** array;
+    size_t size;  /**< Number of keys contained in the KeySet */
+    size_t alloc; /**< Allocated size of array */
+
+    Opmphm * opmphm;
+    OpmphmPredictor * opmphmPredictor;
+
+    uint16_t refs; /**< Reference counter */
+};
+
+struct _KeySet {
+    struct _KeySetData * data;
+
+    struct _Key * cursor; /**< Internal cursor */
+    size_t current;		  /**< Current position of cursor */
+	
+    ksflag_t flags;
+
+    uint16_t refs; /**< Reference counter */
+};
+```
+
+The following calculations are based on the AMD64 platform:
+
+An empty keyset with the datastructure above has 80 bytes.
+An empty keyset with the current implementation has 64 bytes.
+
+A copied, non-modified keyset with the datastructure above has always 32 bytes.
+A copied, non-modified keyset with the current implementation has at least 64 bytes (for an empty keyset).
+
```

Comment:
Sizes are not the only criteria, a disadvantage of this approach is the probably higher number of allocations.

Reply:
> higher number of allocations.

That is an issue I was concerned with too. But do we actually have any data to show that this is a problem? Is it really the high number of individual allocations or is it actually the larger number of indirections?

If the problem is just a large number of small allocations, then we could maybe implement a pool allocator. Then there would just be one (or multiple) bigger pools for each type of struct.

Reply:
What exactly are our concerns with the number of allocations? @markus2330 can you formulate a research question for this? I can then do some research and benchmarking to find the answer for it.

Reply:
> What exactly are our concerns with the number of allocations? @markus2330 can you formulate a research question for this?

The concerns are basically covered with CPU+memory overhead. I was throwing this in as it is easy to calculate during the decision phase. Many more allocations are definitely not favorable for CPU+memory overhead and give a more complete picture than simply looking at bytes.

Ideally you could make a table with the different COW options and their estimated resource usage (bytes+nr of allocations). For the thesis this will be nice; if it will already help for the decision, I have to admit that I don't know.